### ISSUE TITLE
Create a POC to  evaluate 8 byte version id creation with collision retry

### ISSUE THREAD

**Issue by iglesiasbrandon** [storj/storj#6220]
### Description:
Create a POC to evaluate INT4, INT8, 8-byte version ID creation with collision retry. 

We want to understand the performance of a table migration from the current version INT4 column to an INT8 column, as opposed to a table migration from INT4 to BYTES. We also want to understand the performance impact and frequency of collisions. 

### Open Question: 
- Do we ever care about versions being unique across all objects?

### Acceptance Criteria
- Determine if we should use INT4 VS INT8 VS Bytes

**Comment by egonelbre** [comment]
For million objects the size seems roughly linear:
```
int4      37.2 MiB
int8      41.1 MiB
bytes(16) 48.3 MiB
```

Test code:

<details>

```
package main

import (
	"context"
	"fmt"

	"github.com/jackc/pgconn"
	"github.com/jackc/pgx/v4"
)

func main() {
	ctx := context.Background()
	Test(ctx, "postgres://storj:storj-pass@localhost/metabase?sslmode=disable")
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func must(_ pgconn.CommandTag, err error) { check(err) }

func Test(ctx context.Context, dburl string) {
	conn, err := pgx.Connect(ctx, dburl)
	check(err)

	must(conn.Exec(ctx, `DROP TABLE IF EXISTS folder;`))
	must(conn.Exec(ctx, `CREATE TABLE folder (name BYTEA PRIMARY KEY)`))
	must(conn.Exec(ctx, `INSERT INTO folder (name) VALUES ($1)`, []byte("a/b/c/\x00\xFF")))
	must(conn.Exec(ctx, `INSERT INTO folder (name) VALUES ($1)`, []byte("a/y/d/\x00\xFF")))
	must(conn.Exec(ctx, `INSERT INTO folder (name) VALUES ($1)`, []byte("z/b/d/\x00\xFF")))
	must(conn.Exec(ctx, `INSERT INTO folder (name) VALUES ($1)`, []byte("a/x/d/\x00\xFF")))
	must(conn.Exec(ctx, `INSERT INTO folder (name) VALUES ($1)`, []byte("a/b/e/\x00\xFF")))

	rows, err := conn.Query(ctx, `
		SELECT substring(name from 2)
		FROM folder
		--WHERE substring(name)
		--WHERE substring(name from 2)name SIMILAR TO 'a/%/d/%'
	`)
	check(err)
	defer rows.Close()

	for rows.Next() {
		var v []byte
		check(rows.Scan(&v))
		//fmt.Println(strconv.Quote(string(v)))
		fmt.Println(string(v))
	}
	check(rows.Err())
}
```

</details>

**Comment by egonelbre** [comment]
The migration from `INT4` to `INT8` is just a table schema change in Cockroach, no need for a slow migration. The encoding for INT4 and INT8 are the same in the index.

It doesn't look like there's a fast conversion from `INT4` to `BYTES`. I'm not sure how slow this migration will be -- it will need to rebuild the index and add the new field.

**Comment by jenlij** [comment]
- How many version IDs are in the db already? 
- How are the version IDs determined right now?
- When migrating version ID, what should the conversion formula be? (pad zeros, combine known values, etc)

- We probably don't need to worry about collisions across different objects (eg object 1 and object 2 can both have version ID 567)
- We can estimate the collision probabilities
- We can test a migration and check for collisions.

**Comment by egonelbre** [comment]
> How many version IDs are in the db already? (looks like there are many ~1000)

Yeah, there are quite a bit of them. I'm not sure how relevant it's going to be.

> How are the version IDs determined right now? (incremental, starting from 1)

For a new upload the version is set based on `max(objects.version)+1`, on the same object key.

> What's the history of the current object versions?

I'm not quite following the question.

> When migrating version ID, what should the conversion formula be? (pad zeros, combine known values, etc)

pad zeros, because that seems the easiest to implement and I'm not aware of any downsides.

> We probably don't need to worry about collisions across different objects (eg object 1 and object 2 can both have version ID 567)

Yes, multiple objects could have the same version. We need to ensure that's the case anyways due to the null version.

> We can estimate the collision probabilities

I assume yes, but it needs to do some birthday problem into account and the scale of the whole database, not just a single object.

e.g. how likely is that in 30 years an object has a version collision assuming each object gets 1000 versions, and we scale 100x during that time.

> We can test a migration and check for collisions.

Migration won't lead to any collision. If it does, then the migration is buggy.

**Comment by dlamarmorgan** [comment]
@jtolio if you can help with finding the probability of a version collision on a single object, and how many versions it would take before we would need to worry about it?

**Comment by jtolio** [comment]
The probability table on https://en.wikipedia.org/wiki/Birthday_problem#Probability_table says that for 64 bits, we would need 610,000,000 versions before we had a 1% chance of collision, and 5,100,000,000 for a 50% chance.
